### PR TITLE
Dvci 278 error states

### PR DESCRIPTION
**Pull request by nkarip**

### CODE REVIEW

File: src/components/HealthCardDisplay/HealthCardDisplay.js [hunk 1]
```diff
@@ -114,133 +114,185 @@ const HealthCardDisplay = () => {
...
+            />
+            <Typography variant="h4">
+              {text}
+            </Typography>
+          </Box>
+        </Container>
+      </Grid>
+    );
+
+    const BottomBanner = ({
```
Comment (mickohanlon23) [inline src/components/HealthCardDisplay/HealthCardDisplay.js:187]: When I scan the SHCs linked here: https://github.com/dvci/shc-web-verifier/pull/8#pullrequestreview-813727922, I'm brought to this weird error page that also displays the bottom banners:

Screen Shot 2022-04-08 at 10 49 55 AM



whereas, when I scan this card on the main branch, I'm taken to the  normal "Not verified" page:

image

**Review by mickohanlon23 — COMMENTED** [review]
Refactoring looks great! Besides the one odd scenario I pointed out in my comment, all other error scenarios that I tested work as expected.

**Review by sarahmcdougall — COMMENTED** [review]
I compared the changes here to our error documentation, and it looks like we have covered all the different error states that are documented. I looked into the bug that Mick found, and I think we may need to investigate the `useEffect` in the HealthCardDisplay and how it sets the top and bottom banner styles.

File: src/components/HealthCardDisplay/HealthCardDisplay.js (hunk 1 above)
Comment (nkarip) [reply src/components/HealthCardDisplay/HealthCardDisplay.js:235]: Made a bunch of changes, should be fixed now

File: src/components/HealthCardDisplay/HealthCardDisplay.js [hunk 2]
```diff
@@ -111,126 +112,75 @@ const HealthCardDisplay = () => {
   const history = useHistory();
   const { t } = useTranslation();
   const {
-    qrCodes, healthCardVerified, issuerVerified, validPrimarySeries
+    qrError,
+    jws,
```
Comment (sarahmcdougall) [inline src/components/HealthCardDisplay/HealthCardDisplay.js:116]:
```suggestion
    jws,
    resetQrCodes
```

File: src/components/HealthCardDisplay/HealthCardDisplay.js [hunk 3]
```diff
@@ -111,126 +112,75 @@ const HealthCardDisplay = () => {
...
   } = useQrDataContext();
-  const patientData = getPatientData(qrCodes);
 
   const handleScan = () => {
     history.push('qr-scan');
```
Comment (sarahmcdougall) [inline src/components/HealthCardDisplay/HealthCardDisplay.js:120]:
```suggestion
    resetQrCodes();
    history.push('qr-scan');
```

Comment (sarahmcdougall) [reply src/components/HealthCardDisplay/HealthCardDisplay.js:120]: I ran into the following bug: if I scanned a QR that was not a valid SHC (stored UNSUPPORTED_QR_NOT_SHC for the `qrError`), then tried to scan a valid SHC right after, I would reach the invalid SHC error page for the valid SHC. It looks like the `qrError` doesn't get reset between scans, and `resetQrCodes()` never gets called. Adding `resetQrCodes()` to `handleScan()` solved the issue for me.

**Review by sarahmcdougall — CHANGES_REQUESTED** [review]
The new `useReducer()` does a good job of handling our changing data. I noticed that we no longer have multiple re-renders when displaying the SHC after scanning. I left some comments about a bug I found when scanning an unsupported QR followed by a valid SHC.

Also, I tried scanning an unverified QR, which first brought me to the unverified error page, but then it errored out. Is this the intended behavior, or do we want to direct to the error page instead of directly throwing an error for an unverified card?

File: src/components/HealthCardDisplay/HealthCardDisplay.js (hunk 3 above)
Comment (nkarip) [reply src/components/HealthCardDisplay/HealthCardDisplay.js:120]: The problem there is we don't want to clear the QR data until the scan is fully complete - if the person goes to qr-scan and then aborts, there is no data in the provider. If they navigate to the display page then it breaks. We want the display route to show the last scanned data stored in history, and resetQrCodes removes that info.

Comment (nkarip) [reply src/components/HealthCardDisplay/HealthCardDisplay.js:120]: Also, I wasn't able to replicate this error - tried scanning an unsupported card then a valid one and it successfully showed the valid one. Can you test again? And attach the QR codes you used for testing?

Comment (sarahmcdougall) [reply src/components/HealthCardDisplay/HealthCardDisplay.js:120]: When I tested this, I used a non-SHC QR (like [this one](https://en.wikipedia.org/wiki/QR_code#/media/File:QR_code_for_mobile_English_Wikipedia.svg) ) and then scanned the first QR from [our Confluence page](https://confluence.mitre.org/pages/viewpage.action?spaceKey=DVCI&title=QR+Codes+for+Testing+the+Web+Verifier). When scanning the second QR, I was brought to the error screen again.

Comment (nkarip) [reply src/components/HealthCardDisplay/HealthCardDisplay.js:120]: Got it, thanks for clarifying. I added the call to resetQrCodes() but at the point where the QR is actually done scanning, so we don't lose the old value in local storage if the user aborts during the scan. Tested with the codes you used and it seems to be working (fingers crossed).

Comment (sarahmcdougall) [reply src/components/HealthCardDisplay/HealthCardDisplay.js:120]: Sounds good, I tested with your changes and I don't run into the bug anymore.

**Review by sarahmcdougall — APPROVED** [review]

File: src/components/HealthCardDisplay/HealthCardDisplay.js (hunk 2 above)
Comment (mickohanlon23) [reply src/components/HealthCardDisplay/HealthCardDisplay.js:116]: I'm also seeing the React warning for not including `resetQrCodes` in the useEffect dependency array.

**Review by mickohanlon23 — APPROVED** [review]
I like the way the `QRDataProvider` is set up with the reducer. I tested with all of the QR Codes at https://confluence.mitre.org/display/DVCI/QR+Codes+for+Testing+the+Web+Verifier and all of the error handling scenarios behave as expected. A minor React error is still coming up, as Sarah pointed out. Other than that, the updated code looks good, and makes the error handling a lot easier to follow!

**Pull request merged**